src/java.base/share/classes/jdk/internal/misc/TerminatingThreadLocal.java

```diff
@@ -80,7 +79,9 @@ public static void threadTerminated() {
      * @param tl the ThreadLocal to register
      */
     public static void register(TerminatingThreadLocal<?> tl) {
-        REGISTRY.get().add(tl);
+        if (tl != REGISTRY) {
```

Comment:
Would this happen in any well-behaved application? If not, might be better to throw an exception? 🤔

Reply:
It's just a consequence of the registry being a ttl, we either check it here or in set, it's more local here.